### PR TITLE
Expose the communication functions

### PR DESCRIPTION
**Pull request by DaikiMaekawa**

### CODE REVIEW

**Comment by DaikiMaekawa** [comment]
@at-wat PTAL

File: CMakeLists.txt [hunk 1]
```diff
@@ -210,6 +210,17 @@ add_library(ypspur SHARED
   src/yprintf.c
   src/utility.c
 )
+
+add_library(ypspur-comm SHARED
+  src/communication.c
+  src/serial.c
```
Comment (at-wat) [inline CMakeLists.txt]: `serial.c` is better to be removed from exposed library. it has global variable managing serial connection of ypspur-coordinator.
Moving `encode_write` to `src/serial.c` would be good.

File: CMakeLists.txt [hunk 2]
```diff
@@ -210,6 +210,17 @@ add_library(ypspur SHARED
...
+  src/utility.c
+)
+
+target_link_libraries(ypspur-comm
+ ypspur
```
Comment (at-wat) [inline CMakeLists.txt]: ypspur is IPC library connecting user program to ypspur-coordinator.
it's not used in ypspur-comm.

**Comment by DaikiMaekawa** [comment]
@at-wat Address your comments.

File: include/communication.h [hunk 3]
```diff
@@ -26,7 +26,6 @@
 #define COMMUNICATION_END_BYTE 0x0a
 
 int encode(unsigned char *src, int len, unsigned char *dst, int buf_max);
-int decord(unsigned char *src, int len, unsigned char *dst, int buf_max);
-int encode_write(char *data, int len);
+int decode(unsigned char *src, int len, unsigned char *dst, int buf_max);
```
Comment (at-wat) [inline include/communication.h]: `extern "C"` would be needed for C++

Comment (at-wat) [reply include/communication.h]: Would be nice to make `src` const to make it easy to be used from modern code.

**Comment by DaikiMaekawa** [comment]
@at-wat I updated the code. Please take a look at it.

**Comment by at-wat** [comment]
install tree
```
.
├── bin
│   ├── ypspur-coordinator
│   ├── ypspur-free
│   └── ypspur-interpreter
├── include
│   ├── cartesian2d.h
│   ├── formula-calc.h
│   ├── ypparam.h
│   ├── ypspur
│   │   └── communication.h
│   ├── yp-spur.h
│   ├── ypspur.h
│   └── ypspur-md.h
├── lib
│   ├── libcarte2d.a
│   ├── libcarte2d.so -> libcarte2d.so.0
│   ├── libcarte2d.so.0 -> libcarte2d.so.0.0.0
│   ├── libcarte2d.so.0.0.0
│   ├── libformula-calc.a
│   ├── libformula-calc.so -> libformula-calc.so.0
│   ├── libformula-calc.so.0 -> libformula-calc.so.0.0.1
│   ├── libformula-calc.so.0.0.1
│   ├── libypspur.a
│   ├── libypspur-comm.so
│   ├── libypspur-md.a
│   ├── libypspur-md.so -> libypspur-md.so.14
│   ├── libypspur-md.so.14 -> libypspur-md.so.14.2.0
│   ├── libypspur-md.so.14.2.0
│   ├── libypspur.so -> libypspur.so.14
│   ├── libypspur.so.14 -> libypspur.so.14.2.0
│   └── libypspur.so.14.2.0
└── share
    └── ypspur
        ├── cmake
        │   ├── YPSpurTargets.cmake
        │   └── YPSpurTargets-noconfig.cmake
        ├── package.xml
        ├── ypspur-config.cmake
        └── ypspur-config-version.cmake
```

```cpp
#include <ypspur/communication.h>
#include <iostream>

int main()
{
  unsigned char test[] = {0x01, 0x02, 0x03, 0x04, 0x05};
  unsigned char dest[100];
  int len = encode(test, sizeof(test), dest, 100);

  std::cerr << "input: "
    << (int)test[0] << ", "
    << (int)test[1] << ", "
    << (int)test[2] << ", "
    << (int)test[3] << ", "
    << (int)test[4] << std::endl;
  std::cerr << "output size: " << len << std::endl;
  std::cerr << "output: "
    << (int)dest[0] << ", "
    << (int)dest[1] << ", "
    << (int)dest[2] << ", "
    << (int)dest[3] << ", "
    << (int)dest[4] << ", "
    << (int)dest[5] << ", "
    << (int)dest[6] << std::endl;

  unsigned char revert[100];
  int len2 = decode(dest, len, revert, 100);
  std::cerr << "reverted size: " << len2 << std::endl;
  std::cerr << "reverted: "
    << (int)revert[0] << ", "
    << (int)revert[1] << ", "
    << (int)revert[2] << ", "
    << (int)revert[3] << ", "
    << (int)revert[4] << std::endl;
}
```

works as 
```
input: 1, 2, 3, 4, 5
output size: 7
output: 64, 80, 72, 67, 65, 64, 84
reverted size: 5
reverted: 1, 2, 3, 4, 5
```

**Comment by at-wat** [comment]
@DaikiMaekawa could you take a look? http://build.ros.org/job/Mdev__yp-spur__ubuntu_bionic_amd64/17/

**Comment by DaikiMaekawa** [comment]
I don't think the CI error occurred by our changes. Let's see how it goes for a while.

**Pull request merged**